llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,371 @@
+    return false;
+
+  for (const MachineInstr *MI : SubRegDefs.at(ID)) {
+    if (MI->isCopyLike())
+      return true;
+  }
+  return false;
+}
+
+std::map<unsigned, SmallVector<const MachineInstr *, 4>>
+getAllSubRegDefs(const MachineBasicBlock &MBB) {
+  std::map<unsigned, SmallVector<const MachineInstr *, 4>> SubRegDefs;
+  for (const MachineInstr &MI : MBB) {
+    for (const MachineOperand Def : MI.defs()) {
+      unsigned ID = Def.getReg().id();
+
+      if (Def.isUndef())
+        SubRegDefs[ID] = SmallVector<const MachineInstr *, 4>();
+
+      if (SubRegDefs.find(ID) == SubRegDefs.end())
+        continue;
+
+      SubRegDefs[ID].push_back(&MI);
+    }
+  }
+  return SubRegDefs;
+}
+
+bool AIEWawRegRewriter::renameMBBPhysRegs(const MachineBasicBlock *MBB) {
+  bool Modified = false;
+  LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+             dbgs() << "\n");
+  BitVector UsedPhyRegs(TRI->getNumRegs());
+  SmallVector<std::pair<MCRegister, MCRegister>, 8> ReplacementRegs;
+
+  SmallSet<MCPhysReg, 16> ForbiddenPhysRegs = getMBBForbiddenReplacements(MBB);
+  std::map<unsigned, SmallVector<const MachineInstr *, 4>> SubRegDefs =
+      getAllSubRegDefs(*MBB);
+
+  for (const MachineInstr &MI : *MBB) {
+
+    if (isIdentityCopy(MI))
+      continue;
+
+    for (const MachineOperand &MO : MI.defs()) {
+
+      Register Reg = MO.getReg();
+      if (!Reg.isVirtual())
+        continue;
+      if (VRM->hasRequiredPhys(Reg))
+        continue;
+      if (MO.isTied())
+        continue;
+      if (!TRI->isVecOrAccRegClass(*(MRI->getRegClass(Reg))))
+        continue;
+      
+      MCRegister PreSwapPhyReg = VRM->getPhys(Reg).id();
+      if (isWAWCandidate(Reg, MI, UsedPhyRegs, SubRegDefs) &&
+          replaceReg(Reg, ForbiddenPhysRegs)) {
+
+        LLVM_DEBUG(dbgs() << MI);
+        Modified = true;
+        ReplacementRegs.push_back(
+            std::make_pair(PreSwapPhyReg, VRM->getPhys(Reg).id()));
+
+      } else {
+        // keep track of already visted physical and virtual Regs
+        UsedPhyRegs[VRM->getPhys(Reg)] = true;
+      }
+    }
+  }
+  // dump Reassignment
+  LLVM_DEBUG(dbgs() << "WAW Replacement Overview for bb." << MBB->getNumber();
+             dbgs() << "\n"; for (std::pair<MCRegister, MCRegister> &RegPair
+                                  : ReplacementRegs) {
+               dbgs() << printReg(RegPair.first, TRI, 0, MRI) << " -> "
+                      << printReg(RegPair.second, TRI, 0, MRI) << "\n";
+             } dbgs() << "\n");
+
+  return Modified;
+}
+
+bool AIEWawRegRewriter::isWAWCandidate(
+    const Register &Reg, const MachineInstr &MI, const BitVector &UsedPhyRegs,
+    const std::map<unsigned, SmallVector<const MachineInstr *, 4>> &SubRegDefs)
+    const {
+  const bool AlreadyUsedPhysReg = UsedPhyRegs[VRM->getPhys(Reg)];
+  const bool SubRegCopy = hasSubRegCopy(Reg.id(), SubRegDefs);
+
+  return MRI->getRegClass(Reg) && AlreadyUsedPhysReg && !MI.isCopy() &&
```

Comment:
Just a question: when `MRI->getRegClass(Reg)` can return null?

Reply:
nit: maybe you can test the trivial things first and then finalize with:

```
return !hasSubRegCopy(Reg.id(), SubRegDefs);
```

Reply:
The check was unnecessary, since i only work on Virtual registers. 
I also simplified the structure

Reply:
I think it's a good point, we seem to check that `Reg.isVirtual()` before entering this function. How about making this an official pre-condition with e.g. `assert(Reg.isVirtual() && MRI->getRegClass(Reg))`?

Reply:
I also added another assert to the `replaceReg` function